File: cogs/minigames.py
import asyncio
import contextlib
import re

from discord.ext import commands

moleman_loc = {
    "<a:MoleMan:1022972147175526441><:emptyspace:827651824739156030><:emptyspace:827651824739156030>": (
        0
    ),
    "<:emptyspace:827651824739156030><a:MoleMan:1022972147175526441><:emptyspace:827651824739156030>": (
        1
    ),
    "<:emptyspace:827651824739156030><:emptyspace:827651824739156030><a:MoleMan:1022972147175526441>": (
        2
    ),
}

worms_loc = {
    "<:emptyspace:827651824739156030><:Worm:864261394920898600><:Worm:864261394920898600>": (
        0
    ),
    "<:Worm:864261394920898600><:emptyspace:827651824739156030><:Worm:864261394920898600>": (
        1
    ),
    "<:Worm:864261394920898600><:Worm:864261394920898600><:emptyspace:827651824739156030>": (
        2
    ),
}


class Minigames(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.Cog.listener()
    async def on_message_edit(self, before, after):
        if (
            after.channel.id != self.bot.channel_id
            or not self.bot.state
            or (after.interaction and after.interaction.user != self.bot.user)
        ):
            return

        for embed in before.embeds:
            embed = embed.to_dict()
            after_embed = after.embeds[0].to_dict()

            # Moleman
            with contextlib.suppress(KeyError):
                if (
                    "Dodge the Worms!" in embed["description"]
                    and "Mole Man" in after_embed["description"]
                ):
                    self.bot.log("Solved Dodge Worms Minigame", "green")
                elif (
                    "Dodge the Worms!" in embed["description"]
                    and "Dodge the Worms!" not in after_embed["description"]
                ):
                    self.bot.log("Failed Dodge Worms Minigame", "red")

        for embed in after.embeds:
            embed = embed.to_dict()

            # MoleMan
            with contextlib.suppress(KeyError):
                if "Dodge the Worms!" in embed["description"]:
                    moleman = embed["description"].splitlines()[5]
                    for i in reversed(embed["description"].splitlines()):
                        if i not in worms_loc:
                            continue
                        match worms_loc[i]:
                            case 0:
                                if moleman_loc[moleman] == 1:
                                    await self.bot.click(after, 0, 0)
                                elif moleman_loc[moleman] == 2:
                                    await self.bot.click(after, 0, 0)
                                    await asyncio.sleep(0.2)
                                    await self.bot.click(after, 0, 0)
                                break
                            case 1:
                                if moleman_loc[moleman] == 0:
                                    await self.bot.click(after, 0, 1)
                                elif moleman_loc[moleman] == 2:
                                    await self.bot.click(after, 0, 0)
                                break
                            case 2:
                                if moleman_loc[moleman] == 0:
                                    await self.bot.click(after, 0, 1)
                                    await asyncio.sleep(0.2)
                                    await self.bot.click(after, 0, 1)
                                if moleman_loc[moleman] == 1:
                                    await self.bot.click(after, 0, 1)
                                break
                    return
            # Football
            with contextlib.suppress(KeyError):
                if "Hit the ball!" in embed["description"]:
                    self.bot.log("Solving Football Minigame", "yellow")
                    if embed["description"].splitlines()[2] == ":levitate:":
                        await self.bot.click(after, 0, 2, [0, 0])
                    elif (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030>:levitate:"
                    ):
                        await self.bot.click(after, 0, 0, [0, 0])
                    if (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030>"
                        "<:emptyspace:827651824739156030>:levitate:"
                    ):
                        await self.bot.click(after, 0, 1, [0, 0])
                    self.bot.log("Solved Football Minigame", "green")
                    return
            # Basketball
            with contextlib.suppress(KeyError):
                if "Dunk the ball!" in embed["description"]:
                    self.bot.log("Solving Basketball Minigame", "yellow")
                    if (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030>"
                        "<:emptyspace:827651824739156030>:basketball:"
                    ):
                        await self.bot.click(after, 0, 2, [0, 0])
                    elif (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030>:basketball:"
                    ):
                        await self.bot.click(after, 0, 1)
                    elif embed["description"].splitlines()[2] == ":basketball:":
                        await self.bot.click(after, 0, 0, [0, 0])
                    self.bot.log("Solved Basketball Minigame", "green")
                    return
            # Dragon
            with contextlib.suppress(KeyError):
                if "Dodge the Fireball" in embed["description"]:
                    self.bot.log("Solving Dragon Minigame", "yellow")
                    if (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030>"
                        "<:emptyspace:827651824739156030><:FireBall:883714770748964864>"
                    ):
                        await self.bot.click(after, 0, 1)
                    elif (
                        embed["description"].splitlines()[2]
                        == "<:emptyspace:827651824739156030><:FireBall:883714770748964864>"
                    ):
                        await self.bot.click(after, 0, 0)
                    elif (
                        embed["description"].splitlines()[2]
                        == "<:FireBall:883714770748964864>"
                    ):
                        await self.bot.click(after, 0, 2)
                    self.bot.log("Solved Dragon Minigame", "green")
                    return
            # Catch the fish
            with contextlib.suppress(KeyError):
                if "Catch the fish!" in embed["description"]:
                    self.bot.log("Solving Fish Minigame", "yellow")
                    if (
                        embed["description"].splitlines()[1]
                        == "<:emptyspace:827651824739156030>"
                        "<:emptyspace:827651824739156030>"
                        "<a:LegendaryFish:971430841211322408>"
                    ):
                        await self.bot.click(after, 0, 2)
                    elif (
                        embed["description"].splitlines()[1]
                        == "<:emptyspace:827651824739156030>"
                        "<a:LegendaryFish:971430841211322408>"
                    ):
                        await self.bot.click(after, 0, 1)
                    elif (
                        embed["description"].splitlines()[1]
                        == "<a:LegendaryFish:971430841211322408>"
                    ):
                        await self.bot.click(after, 0, 0)
                    if (
                        embed["description"].splitlines()[1]
                        == "<:emptyspace:827651824739156030>"
                        "<:emptyspace:827651824739156030><:Kraken:860228238956429313>"
                    ):
                        await self.bot.click(after, 0, 2)
                    elif (
                        embed["description"].splitlines()[1]
                        == "<:emptyspace:827651824739156030><:Kraken:860228238956429313>"
                    ):
                        await self.bot.click(after, 0, 1)
                    elif (
                        embed["description"].splitlines()[1]
                        == "<:Kraken:860228238956429313>"
                    ):
                        await self.bot.click(after, 0, 0)
                    self.bot.log("Solved Fish Minigame", "green")
                    return

    @commands.Cog.listener()
    async def on_message(self, message):
        if (
            message.channel.id != self.bot.channel_id
            or not self.bot.state
            or (message.interaction and message.interaction.user != self.bot.user)
        ):
            return

        for embed in message.embeds:
            embed = embed.to_dict()
            with contextlib.suppress(KeyError):
                if "Dodge the Worms!" in embed["description"]:
                    self.bot.log("Solving Dodge Worms Minigame", "yellow")

            # Color match
            with contextlib.suppress(KeyError):
                if (
                    "Look at each color next to the words closely!"
                    in embed["description"]
                ):
                    self.bot.log("Solving Color Match Minigame", "yellow")
                    self.bot.last_ran = {
                        k: v + 100 if v != 0 else float("inf")
                        for k, v in self.bot.last_ran.items()
                    }
                    options = {}
                    for line in embed["description"].splitlines()[1:]:
                        match_word = re.search("`(.+?)`", line)
                        match_color = re.search(":([^:]+):", line)
                        if match_word and match_color:
                            options[match_word[1]] = match_color[1]
                    await asyncio.sleep(6)
                    embed = message.embeds[0].to_dict()
                    word = re.search("`(.+?)`", embed["description"])[1]
                    color = options[word]
                    for count, button in enumerate(message.components[0].children):
                        if button.label == color:
                            await self.bot.click(message, 0, count)
                    self.bot.last_ran = {
                        k: v - 100 if v != float("inf") else 0
                        for k, v in self.bot.last_ran.items()
                    }
                    self.bot.log("Solved Color Match Minigame", "green")
                    return

            # Emoji
            with contextlib.suppress(KeyError):
                if "Look at the emoji closely!" in embed["description"]:
                    self.bot.log("Solving Emoji Minigame", "yellow")
                    self.bot.last_ran = {
                        k: v + 100 if v != 0 else float("inf")
                        for k, v in self.bot.last_ran.items()
                    }
                    emoji = str(embed["description"].splitlines()[1])
                    await asyncio.sleep(4)
                    for row, i in enumerate(message.components):
                        for column, button in enumerate(i.children):
                            if str(button.emoji) == emoji:
                                await self.bot.click(message, row, column)
                    self.bot.last_ran = {
                        k: v - 100 if v != float("inf") else 0
                        for k, v in self.bot.last_ran.items()
                    }
                    self.bot.log("Solved Emoji Minigame", "green")
                    return
            # Repeat order
            with contextlib.suppress(KeyError):
                if any(
                    i in embed["description"]
                    for i in ["Repeat Order", "word order.", "words order"]
                ):
                    self.bot.log("Solving Repeat Order Minigame", "yellow")
                    self.bot.last_ran = {
                        k: v + 100 if v != 0 else float("inf")
                        for k, v in self.bot.last_ran.items()
                    }
                    order = [
                        line[1:-1]
                        for line in message.embeds[0].description.splitlines()[1:6]
                    ]
                    await asyncio.sleep(6)
                    answers = {
                        button.label: i
                        for i, button in enumerate(message.components[0].children)
                    }
                    for choice in order:
                        await self.bot.click(message, 0, answers[choice])
                    self.bot.last_ran = {
                        k: v - 100 if v != float("inf") else 0
                        for k, v in self.bot.last_ran.items()
                    }
                    self.bot.log("Solved Repeat Order Minigame", "green")
                    return
            # Attack boss
            with contextlib.suppress(KeyError):
                if "Attack the boss by clicking" in embed["description"]:
                    self.bot.log("Solving Attack Boss Minigame", "yellow")
                    self.bot.last_ran = {
                        k: v + 100 if v != 0 else float("inf")
                        for k, v in self.bot.last_ran.items()
                    }
                    while not message.components[0].children[0].disabled:
                        await self.bot.click(message, 0, 0)
                    self.bot.last_ran = {
                        k: v - 100 if v != float("inf") else 0
                        for k, v in self.bot.last_ran.items()
                    }
                    self.bot.log("Solved Attack Boss Minigame", "green")
                    return
            # F in the chat
            with contextlib.suppress(KeyError):
                if embed["description"] == "F":
                    self.bot.log("Solving F In The Chat Minigame", "yellow")
                    await self.bot.click(message, 0, 0)
                    self.bot.log("Solved F In The Chat Minigame", "green")
                    return
            # HighLow
            with contextlib.suppress(KeyError):
                if "I just chose a secret number" in embed["description"]:
                    num = int(
                        re.search("\*\*(.*?)\*\*", embed["description"])[1].title()
                    )
                    if num >= 50:
                        await self.bot.click(message, 0, 0)
                    else:
                        await self.bot.click(message, 0, 2)
                    return


async def setup(bot):
    await bot.add_cog(Minigames(bot))

File: instance/minigames.go
package instance

import (
	"fmt"
	"github.com/BridgeSenseDev/Dank-Memer-Grinder/gateway"
	"regexp"
	"strconv"
	"strings"

	"github.com/BridgeSenseDev/Dank-Memer-Grinder/discord/types"
)

var (
	emoji                       = ""
	colorMatchOptions           = make(map[string]string)
	repeatOrder                 []string
	repeatOrderLastClickedIndex int
	wordRegex                   = regexp.MustCompile(`(.+?)`)
	colorRegex                  = regexp.MustCompile(`:([^:]+):`)
	highlowRegex                = regexp.MustCompile(`\*\*(.*?)\*\*`)
)

const (
	emptyspace        = "<:emptyspace:827651824739156030>"
	levitate          = ":levitate:"
	basketball        = ":basketball:"
	fireBall          = "<:FireBall:883714770748964864>"
	worm              = "<:Worm:864261394920898600>"
	PinkBits          = "<:PinkBits:975398146152738906>"
	PinkSludgeMonster = "<:PinkSludgeMonster:1127334051422937240>"
)

func generateEmojiActions(emojis []string) map[string]int {
	emojiActions := make(map[string]int)
	for _, emoji := range emojis {
		if emoji == levitate {
			// Miss the goalkeeper
			emojiActions[emoji] = 2
			emojiActions[emptyspace+emoji] = 2
			emojiActions[emptyspace+emptyspace+emoji] = 0
		} else if emoji == fireBall || emoji == worm || emoji == PinkBits {
			// Dodge the fireball / dodge the worm // dodge the pink bits
			emojiActions["# "+emptyspace+emoji] = 2
			emojiActions["# "+emptyspace+emptyspace+emoji] = 2
			emojiActions["# "+
				emptyspace+emptyspace+emptyspace+emoji] = 0
		} else {
			emojiActions[emoji] = 0
			emojiActions[emptyspace+emoji] = 1
			emojiActions[emptyspace+emptyspace+emoji] = 2
		}
	}
	return emojiActions
}

func (in *Instance) solveMinigame(message gateway.EventMessage, gameName string, gameTrigger string, emojis []string) {
	embed := message.Embeds[0]

	if strings.Contains(embed.Description, gameTrigger) {
		in.Log("others", "INF", fmt.Sprintf("Solving %s minigame", gameName))
		gameActions := generateEmojiActions(emojis)
		for emoji, button := range gameActions {
			if strings.Split(embed.Description, "\n")[2] == emoji {
				err := in.ClickButton(message, 0, button)
				if err != nil {
					in.Log("discord", "ERR", fmt.Sprintf("Failed to click %s minigame button", gameName))
				}
				break
			}
		}
		in.Log("others", "INF", fmt.Sprintf("Solved %s minigame", gameName))
	}
}

func (in *Instance) MinigamesMessageCreate(message gateway.EventMessage) {
	embed := message.Embeds[0]

	// Dragon
	in.solveMinigame(message, "Dragon", "Dodge the Dragon's Fireball", []string{fireBall})

	// Pink sludge monster
	in.solveMinigame(message, "Pink Sludge Monster", "Dodge the Sludge Monster's Pink Bits", []string{PinkBits})

	// Moleman
	in.solveMinigame(message, "Moleman", "Dodge the Moleman's Worm", []string{worm})

	// Attack boss
	if strings.Contains(embed.Description, "Attack the boss by clicking") {
		in.Log("others", "INF", "Solving attack boss minigame")
		in.PauseCommands(false)
		err := in.ClickButton(message, 0, 0)
		if err != nil {
			in.Log("discord", "ERR", fmt.Sprintf("Failed to click attack boss minigame button: %s", err.Error()))
		}
		return
	}

	// Emoji match
	if strings.Contains(embed.Description, "Look at the emoji closely!") {
		in.PauseCommands(false)
		in.Log("others", "INF", "Solving emoji minigame")

		lines := strings.Split(embed.Description, "\n")
		emoji = lines[1]
	}

	// Color match
	if strings.Contains(embed.Description, "Look at each color next to the words closely!") {
		in.PauseCommands(false)
		in.Log("others", "INF", "Solving color match minigame")

		lines := strings.Split(embed.Description, "\n")[1:]

		for _, line := range lines {
			matchWord := wordRegex.FindStringSubmatch(line)
			matchColor := colorRegex.FindStringSubmatch(line)

			if len(matchWord) > 1 && len(matchColor) > 1 {
				colorMatchOptions[matchWord[1]] = matchColor[1]
			}
		}
	}

	// Repeat order
	if strings.Contains(embed.Description, "Repeat Order") ||
		strings.Contains(embed.Description, "word order.") ||
		strings.Contains(embed.Description, "words order") {
		in.Log("others", "INF", "Solving repeat order minigame")
		in.PauseCommands(false)

		lines := strings.Split(embed.Description, "\n")[1:6]
		repeatOrder = make([]string, len(lines))
		for i, line := range lines {
			repeatOrder[i] = strings.Trim(line, "`")
		}
		repeatOrderLastClickedIndex = 0
	}

	// F in the chat
	if embed.Description == "F" {
		err := in.ClickButton(message, 0, 0)
		if err != nil {
			in.Log("discord", "ERR", fmt.Sprintf("Failed to click F in the chat minigame button: %s", err.Error()))
		}
		in.Log("others", "INF", "Solved F in the chat minigame")
	}

	// HighLow
	if strings.Contains(embed.Description, "I just chose a secret number") && message.Interaction.Name != "highlow" {
		matches := highlowRegex.FindStringSubmatch(embed.Description)
		if len(matches) > 1 {
			numStr := matches[1]
			num, _ := strconv.Atoi(numStr)

			if num >= 50 {
				err := in.ClickButton(message, 0, 0)
				if err != nil {
					in.Log("discord", "ERR", fmt.Sprintf("Failed to click highlow minigame button: %s", err.Error()))
				}
			} else {
				err := in.ClickButton(message, 0, 2)
				if err != nil {
					in.Log("discord", "ERR", fmt.Sprintf("Failed to click highlow minigame button: %s", err.Error()))
				}
			}

			in.Log("others", "INF", "Solved highlow minigame")
		} else {
			in.Log("others", "ERR", "Failed to solve highlow minigame")
		}
	}
}

func (in *Instance) MinigamesMessageUpdate(message gateway.EventMessage) {
	embed := message.Embeds[0]

	// Football
	in.solveMinigame(message, "Football", "Hit the ball!", []string{levitate})

	// Basketball
	in.solveMinigame(message, "Basketball", "Dunk the ball!", []string{basketball})

	// Attack boss
	if strings.Contains(embed.Description, "Attack the boss by clicking") {
		if !message.Components[0].(*types.ActionsRow).Components[0].(*types.Button).Disabled {
			err := in.ClickButton(message, 0, 0)
			if err != nil {
				in.Log("discord", "ERR", fmt.Sprintf("Failed to click attack boss minigame button: %s", err.Error()))
			}
		} else {
			in.Log("others", "INF", "Solved attack boss minigame")
			in.UnpauseCommands()
			return
		}
	}

	// Emoji match
	if strings.Contains(embed.Description, "What was the emoji?") {
		for rowIndex, row := range message.Components {
			for columnIndex, button := range row.(*types.ActionsRow).Components {
				if button.(*types.Button).Emoji.Name == emoji {
					err := in.ClickButton(message, rowIndex, columnIndex)
					if err != nil {
						in.Log("discord", "ERR", fmt.Sprintf("Failed to click emoji match minigame button: %s", err.Error()))
					}
					in.Log("others", "INF", "Solved emoji minigame")
					in.UnpauseCommands()
				}
			}
		}
	}

	// Color match
	if strings.Contains(embed.Description, "What color was next to the word") {
		matches := wordRegex.FindStringSubmatch(embed.Description)
		if len(matches) > 1 {
			word := matches[1]
			color := colorMatchOptions[word]

			for columnIndex, button := range message.Components[0].(*types.ActionsRow).Components {
				if button.(*types.Button).Label == color {
					err := in.ClickButton(message, 0, columnIndex)
					if err != nil {
						in.Log("discord", "ERR", fmt.Sprintf("Failed to click color match minigame button: %s", err.Error()))
					}
					break
				}
			}

			in.UnpauseCommands()
			in.Log("others", "INF", "Solved color match minigame")
		} else {
			in.Log("others", "ERR", "Failed to solve color match minigame")
		}
	}

	// Repeat order
	if strings.Contains(embed.Description, "Click the buttons in correct order") {
		answers := make(map[string]int)
		for i, button := range message.Components[0].(*types.ActionsRow).Components {
			answers[button.(*types.Button).Label] = i
		}

		if repeatOrderLastClickedIndex < len(repeatOrder) {
			choice := repeatOrder[repeatOrderLastClickedIndex]
			err := in.ClickButton(message, 0, answers[choice])
			if err != nil {
				in.Log("discord", "ERR", fmt.Sprintf("Failed to click repeat order minigame button: %s", err.Error()))
			} else {
				repeatOrderLastClickedIndex++
				if repeatOrderLastClickedIndex >= len(repeatOrder) {
					in.Log("others", "INF", "Solved repeat order minigame")
					in.UnpauseCommands()
				}
			}
		}
	}
}
